In [ ]:
import pandas as pd

# Load data from a CSV file
df = pd.read_csv('/content/GA_2_dataset.csv')

# Display the first few rows of the DataFrame
print(df.head())

   PlayerID   Age  Gender Location   GameGenre  PlayTimeHours  \
0     35900  37.0    Male    Other    Strategy      23.929404   
1     27085  25.0    Male      NaN      Action      22.755168   
2     39595  24.0  Female   Europe  Simulation      19.505292   
3     37440  26.0  Female   Europe         RPG      11.009645   
4     22882  17.0  Female      USA         RPG       0.581039   

   InGamePurchases GameDifficulty  SessionsPerWeek  AvgSessionDurationMinutes  \
0              NaN           Hard                3                        124   
1              1.0           Easy               14                         84   
2              0.0           Hard                3                        172   
3              NaN            NaN                3                         83   
4              1.0         Medium                5                        163   

   PlayerLevel  AchievementsUnlocked EngagementLevel  
0           99                    18          Medium  
1           

In [ ]:
df.shape

(10000, 13)

In [ ]:
print(df.select_dtypes(include='object').columns.tolist())

['Gender', 'Location', 'GameGenre', 'GameDifficulty', 'EngagementLevel']


In [ ]:
# Filter the DataFrame for 'Males' from 'Europe' who made 'InGamePurchases'
filtered_df = df[(df['Gender'] == 'Male') &
                 (df['Location'] == 'Europe') &
                 (df['InGamePurchases'] == 1)]

# Count the number of rows in the filtered DataFrame
num_players = len(filtered_df)

print(f"Number of Males from Europe who made InGamePurchases: {num_players}")

Number of Males from Europe who made InGamePurchases: 299


In [ ]:
# Filter the DataFrame for players under 18 and with PlayTimeHours strictly greater than 10
filtered_players = df[(df['Age'] < 18) & (df['PlayTimeHours'] > 10)]

# Count the number of players in the filtered DataFrame
num_filtered_players = len(filtered_players)

print(f"Number of players under age 18 with strictly greater than 10 PlayTimeHours: {num_filtered_players}")

Number of players under age 18 with strictly greater than 10 PlayTimeHours: 453


In [ ]:
# Define the target variable (y)
y = df['EngagementLevel']

# Define the feature matrix (X) by dropping the target column
X = df.drop('EngagementLevel', axis=1)

# Calculate the total number of null values in the entire dataset
total_null_values = df.isnull().sum().sum()

print(f"Total null values in the dataset: {total_null_values}")

Total null values in the dataset: 3337


In [ ]:
from sklearn.model_selection import train_test_split

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Get value counts for y_train
y_train_counts = y_train.value_counts()

# Find the category with the least value count
least_common_category = y_train_counts.idxmin()

print(f"Category with the least value counts in y_train: {least_common_category}")

Category with the least value counts in y_train: High


# Task
Impute missing values in the 'Age' column of X_train and X_test using the mean of 'Age' from X_train, then print the sum of the imputed 'Age' column from X_test. Impute missing values in the 'Location' column of both X_train and X_test with the constant 'Other'. Impute missing values in the 'GameDifficulty' column of X_train and X_test using the mode of 'GameDifficulty' from X_train. Impute missing values in the 'InGamePurchases' column of both X_train and X_test with the constant 0.

## Impute 'Age' with Mean

### Subtask:
Calculate the mean of the 'Age' column from X_train and use it to fill missing values in both X_train['Age'] and X_test['Age']. Then, calculate and print the sum of the imputed 'Age' column from X_test.


**Reasoning**:
First, I'll calculate the mean of the 'Age' column from the `X_train` DataFrame, then use this mean to fill missing values in both the `X_train` and `X_test` 'Age' columns. Finally, I'll calculate and print the sum of the imputed 'Age' column in `X_test`.



In [ ]:
import numpy as np

# 1. Calculate the mean of the 'Age' column from X_train
mean_age = X_train['Age'].mean()

# 2. Fill the missing values in the 'Age' column of X_train with the calculated mean_age
X_train['Age'].fillna(mean_age, inplace=True)

# 3. Fill the missing values in the 'Age' column of X_test with the calculated mean_age
X_test['Age'].fillna(mean_age, inplace=True)

# 4. Calculate the sum of the 'Age' column in the X_test DataFrame after imputation
sum_imputed_age_X_test = X_test['Age'].sum()

# 5. Print the calculated sum of the 'Age' column from X_test
print(f"Sum of the imputed 'Age' column in X_test: {sum_imputed_age_X_test}")

Sum of the imputed 'Age' column in X_test: 63585.239130434784


/tmp/ipython-input-2103831770.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  X_train['Age'].fillna(mean_age, inplace=True)
/tmp/ipython-input-2103831770.py:10: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try usin

**Reasoning**:
The previous code produced a FutureWarning regarding the use of `inplace=True` with `fillna` on a chained assignment. To address this, I will explicitly reassign the result of the `fillna` operation back to the DataFrame column, which is the recommended practice for modifying DataFrames.



In [ ]:
import numpy as np

# 1. Calculate the mean of the 'Age' column from X_train
mean_age = X_train['Age'].mean()

# 2. Fill the missing values in the 'Age' column of X_train with the calculated mean_age
X_train['Age'] = X_train['Age'].fillna(mean_age)

# 3. Fill the missing values in the 'Age' column of X_test with the calculated mean_age
X_test['Age'] = X_test['Age'].fillna(mean_age)

# 4. Calculate the sum of the 'Age' column in the X_test DataFrame after imputation
sum_imputed_age_X_test = X_test['Age'].sum()

# 5. Print the calculated sum of the 'Age' column from X_test
print(f"Sum of the imputed 'Age' column in X_test: {sum_imputed_age_X_test}")

Sum of the imputed 'Age' column in X_test: 63585.239130434784


## Impute 'Location' with Constant 'Other'

### Subtask:
Impute missing values in the 'Location' column of both `X_train` and `X_test` with the constant string 'Other'. Print the count of 'Other' in `X_train['Location']` after imputation.

**Reasoning**:
Now I will perform the imputation of missing 'Location' values in both X_train and X_test with the constant 'Other', and then print the count of 'Other' in X_train['Location'] to verify.



In [ ]:
X_train['Location'] = X_train['Location'].fillna('Other')
X_test['Location'] = X_test['Location'].fillna('Other')

other_count = (X_train['Location'] == 'Other').sum()
print(f"Count of 'Other' in X_train['Location'] after imputation: {other_count}")

Count of 'Other' in X_train['Location'] after imputation: 1344


## Impute 'GameDifficulty' with Mode

### Subtask:
Impute missing values in the 'GameDifficulty' column of both `X_train` and `X_test` using the mode of 'GameDifficulty' from `X_train`. Print the mode of `X_train['GameDifficulty']` that was used for imputation, and then print the count of missing values in `X_test['GameDifficulty']` after imputation.

**Reasoning**:
Now I will perform the imputation of missing 'GameDifficulty' values in both X_train and X_test using the mode from X_train. I will then print the mode used and verify the imputation by checking the count of missing values in X_test['GameDifficulty'].



In [ ]:
mode_game_difficulty = X_train['GameDifficulty'].mode()[0]

X_train['GameDifficulty'] = X_train['GameDifficulty'].fillna(mode_game_difficulty)
X_test['GameDifficulty'] = X_test['GameDifficulty'].fillna(mode_game_difficulty)

print(f"Mode of 'GameDifficulty' used for imputation: {mode_game_difficulty}")
missing_in_X_test = X_test['GameDifficulty'].isnull().sum()
print(f"Count of missing values in X_test['GameDifficulty'] after imputation: {missing_in_X_test}")

Mode of 'GameDifficulty' used for imputation: Easy
Count of missing values in X_test['GameDifficulty'] after imputation: 0


## Impute 'InGamePurchases' with Constant 0

### Subtask:
Impute missing values in the 'InGamePurchases' column of both `X_train` and `X_test` with the constant value 0. Print the sum of the 'InGamePurchases' column from `X_test` after imputation.

**Reasoning**:
Now I will impute missing values in the 'InGamePurchases' column of both X_train and X_test with the constant 0, and then print the sum of the 'InGamePurchases' column from X_test after imputation.



In [ ]:
X_train['InGamePurchases'] = X_train['InGamePurchases'].fillna(0)
X_test['InGamePurchases'] = X_test['InGamePurchases'].fillna(0)

sum_ingame_purchases_X_test = X_test['InGamePurchases'].sum()
print(f"Sum of 'InGamePurchases' in X_test after imputation: {sum_ingame_purchases_X_test}")

Sum of 'InGamePurchases' in X_test after imputation: 409.0


## Summary:

### Data Analysis Key Findings

*   Missing values in the 'Age' column were imputed using the mean from the training set, resulting in a sum of `63585.239` for the 'Age' column in `X_test` after imputation.
*   The 'Location' column's missing values were filled with the constant 'Other', leading to `1344` occurrences of 'Other' in `X_train['Location']` post-imputation.
*   The mode of 'GameDifficulty' from `X_train`, which was 'Easy', was used to impute missing values in this column for both datasets. This successfully eliminated all `0` missing values in `X_test['GameDifficulty']` after imputation.
*   Missing values in the 'InGamePurchases' column were imputed with the constant `0`, and the sum of this column in `X_test` after imputation was `409.0`.

### Insights or Next Steps

*   The chosen imputation strategies (mean for numerical, mode for categorical, and constant for specific cases) align with common practices for handling missing data, ensuring data completeness without introducing significant bias or noise for these specific features.
*   The dataset is now prepared for further preprocessing steps, such as one-hot encoding for categorical features like 'Location' and 'GameDifficulty', and scaling for numerical features like 'Age' and 'InGamePurchases', before model training.


# Task
Drop the 'PlayerID' column from both X_train and X_test.

## Drop PlayerID Column

### Subtask:
Drop the 'PlayerID' column from both X_train and X_test datasets, as it is not a feature.


**Reasoning**:
To complete the subtask, I will drop the 'PlayerID' column from both X_train and X_test DataFrames, as it is an identifier and not a predictive feature.



In [ ]:
X_train.drop('PlayerID', axis=1, inplace=True)
X_test.drop('PlayerID', axis=1, inplace=True)

print("PlayerID column dropped from X_train and X_test.")

PlayerID column dropped from X_train and X_test.


## Ordinal Encode GameDifficulty

### Subtask:
Apply ordinal encoding to the 'GameDifficulty' column in both X_train and X_test using the specified order: Easy (0), Medium (1), Hard (2).


**Reasoning**:
To apply ordinal encoding, I need to import `OrdinalEncoder` from `sklearn.preprocessing`, define the desired order for the categories, and then instantiate, fit, and transform the 'GameDifficulty' column in both X_train and X_test.



In [ ]:
from sklearn.preprocessing import OrdinalEncoder

# 2. Define the categories for 'GameDifficulty'
categories_difficulty = [['Easy', 'Medium', 'Hard']]

# 3. Instantiate an OrdinalEncoder object
encoder = OrdinalEncoder(categories=categories_difficulty, handle_unknown='use_encoded_value', unknown_value=-1)

# 4. Fit the OrdinalEncoder on the 'GameDifficulty' column of X_train
encoder.fit(X_train[['GameDifficulty']])

# 5. Transform the 'GameDifficulty' column in both X_train and X_test
X_train['GameDifficulty'] = encoder.transform(X_train[['GameDifficulty']])
X_test['GameDifficulty'] = encoder.transform(X_test[['GameDifficulty']])

# 6. Print the unique values of the 'GameDifficulty' column from X_train after encoding
print(f"Unique values in X_train['GameDifficulty'] after encoding: {X_train['GameDifficulty'].unique()}")

Unique values in X_train['GameDifficulty'] after encoding: [1. 0. 2.]


## One-Hot Encode Nominal Features

### Subtask:
Apply one-hot encoding to 'Gender', 'Location', and 'GameGenre' columns in both X_train and X_test, ensuring drop_first=True.


**Reasoning**:
To apply one-hot encoding, I will import `OneHotEncoder`, define the columns to be encoded, instantiate the encoder with specified parameters, then fit and transform the data, create new DataFrames, drop the original columns, and concatenate the new one-hot encoded DataFrames with the respective `X_train` and `X_test` DataFrames. Finally, I will print the first five rows of `X_train`.



In [ ]:
from sklearn.preprocessing import OneHotEncoder
import pandas as pd

# Define nominal categorical columns to be encoded
nominal_cols = ['Gender', 'Location', 'GameGenre']

# Instantiate OneHotEncoder
encoder = OneHotEncoder(handle_unknown='ignore', sparse_output=False, drop_first=True)

# Fit the encoder on the training data for selected nominal columns
encoder.fit(X_train[nominal_cols])

# Transform both X_train and X_test for these columns
X_train_encoded = encoder.transform(X_train[nominal_cols])
X_test_encoded = encoder.transform(X_test[nominal_cols])

# Create DataFrame with appropriate column names
encoded_feature_names = encoder.get_feature_names_out(nominal_cols)
X_train_encoded_df = pd.DataFrame(X_train_encoded, columns=encoded_feature_names, index=X_train.index)
X_test_encoded_df = pd.DataFrame(X_test_encoded, columns=encoded_feature_names, index=X_test.index)

# Drop original nominal columns from X_train and X_test
X_train = X_train.drop(nominal_cols, axis=1)
X_test = X_test.drop(nominal_cols, axis=1)

# Concatenate the new one-hot encoded DataFrames with the respective X_train and X_test DataFrames
X_train = pd.concat([X_train, X_train_encoded_df], axis=1)
X_test = pd.concat([X_test, X_test_encoded_df], axis=1)

# Print the first five rows of the X_train DataFrame to display the updated features
print(X_train.head())

TypeError: OneHotEncoder.__init__() got an unexpected keyword argument 'drop_first'

In [ ]:
from sklearn.preprocessing import StandardScaler

# Instantiate StandardScaler
scaler = StandardScaler()

# Fit on X_train and transform both X_train and X_test
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Convert scaled arrays back to DataFrames
X_train_scaled = pd.DataFrame(X_train_scaled, columns=X_train.columns, index=X_train.index)
X_test_scaled = pd.DataFrame(X_test_scaled, columns=X_test.columns, index=X_test.index)

# Calculate the sum of all values in the first five rows of the transformed test feature matrix
sum_first_five_rows_X_test_scaled = X_test_scaled.head(5).sum().sum()

print(f"Sum of all values in the first five rows of the transformed test feature matrix: {sum_first_five_rows_X_test_scaled:.2f}")

Sum of all values in the first five rows of the transformed test feature matrix: -7.17


**Reasoning**:
The error `TypeError: OneHotEncoder.__init__() got an unexpected keyword argument 'drop_first'` indicates that the `drop_first` parameter is not a valid argument for `OneHotEncoder` in the current scikit-learn version. To achieve the intended functionality of dropping the first category to avoid multicollinearity, the correct parameter to use is `drop='first'`.



In [ ]:
from sklearn.preprocessing import OneHotEncoder
import pandas as pd

# Define nominal categorical columns to be encoded
nominal_cols = ['Gender', 'Location', 'GameGenre']

# Instantiate OneHotEncoder with drop='first'
encoder = OneHotEncoder(handle_unknown='ignore', sparse_output=False, drop='first')

# Fit the encoder on the training data for selected nominal columns
encoder.fit(X_train[nominal_cols])

# Transform both X_train and X_test for these columns
X_train_encoded = encoder.transform(X_train[nominal_cols])
X_test_encoded = encoder.transform(X_test[nominal_cols])

# Create DataFrame with appropriate column names
encoded_feature_names = encoder.get_feature_names_out(nominal_cols)
X_train_encoded_df = pd.DataFrame(X_train_encoded, columns=encoded_feature_names, index=X_train.index)
X_test_encoded_df = pd.DataFrame(X_test_encoded, columns=encoded_feature_names, index=X_test.index)

# Drop original nominal columns from X_train and X_test
X_train = X_train.drop(nominal_cols, axis=1)
X_test = X_test.drop(nominal_cols, axis=1)

# Concatenate the new one-hot encoded DataFrames with the respective X_train and X_test DataFrames
X_train = pd.concat([X_train, X_train_encoded_df], axis=1)
X_test = pd.concat([X_test, X_test_encoded_df], axis=1)

# Print the first five rows of the X_train DataFrame to display the updated features
print(X_train.head())

       Age  PlayTimeHours  InGamePurchases  GameDifficulty  SessionsPerWeek  \
9254  34.0      15.777862              0.0             1.0                5   
1561  41.0       3.044182              0.0             1.0                7   
1670  48.0       0.585613              0.0             1.0               16   
6087  19.0       6.194149              0.0             0.0                6   
6669  38.0      21.610488              0.0             0.0               14   

      AvgSessionDurationMinutes  PlayerLevel  AchievementsUnlocked  \
9254                         76           80                    33   
1561                         73           34                     5   
1670                        163           44                     3   
6087                         87           26                    33   
6669                        116           39                    14   

      Gender_Male  Location_Europe  Location_Other  Location_USA  \
9254          0.0              0.0  

## Scale All Features

### Subtask:
Apply StandardScaler to all numerical and transformed categorical features in both X_train and X_test datasets. After scaling, calculate the sum of all values in the first five rows of the transformed test feature matrix.


## Summary:

### Q&A

*   **Was the 'PlayerID' column successfully dropped from `X_train` and `X_test`?**
    Yes, the 'PlayerID' column was successfully dropped from both `X_train` and `X_test`.
*   **Was the 'GameDifficulty' column ordinally encoded in `X_train` and `X_test`?**
    Yes, the 'GameDifficulty' column was successfully ordinal encoded in both `X_train` and `X_test` using the specified order (Easy: 0, Medium: 1, Hard: 2). The unique values in `X_train['GameDifficulty']` after encoding were `[1. 0. 2.]`.
*   **Were 'Gender', 'Location', and 'GameGenre' columns one-hot encoded in `X_train` and `X_test` with `drop_first=True`?**
    Yes, the 'Gender', 'Location', and 'GameGenre' columns were successfully one-hot encoded in both `X_train` and `X_test` using `drop='first'`. The original columns were removed, and the new encoded features were integrated into the DataFrames.

### Data Analysis Key Findings

*   The 'PlayerID' column was successfully removed from both `X_train` and `X_test` datasets, confirming its identification as a non-feature identifier.
*   The 'GameDifficulty' column was transformed using ordinal encoding. The categories 'Easy', 'Medium', and 'Hard' were mapped to 0, 1, and 2 respectively, as evidenced by the unique encoded values `[1. 0. 2.]` in `X_train`.
*   Nominal categorical features ('Gender', 'Location', 'GameGenre') were successfully one-hot encoded. An initial attempt to use `drop_first=True` with `OneHotEncoder` resulted in a `TypeError`, which was resolved by using the correct parameter `drop='first'`. This ensured that the first category of each feature was dropped to avoid multicollinearity.
*   The original nominal columns were replaced with their one-hot encoded counterparts, and the updated `X_train` and `X_test` DataFrames now contain these new features.

### Insights or Next Steps

*   All specified categorical and identifier features have been preprocessed. The datasets are now prepared for the next step, which involves scaling the numerical features.
*   The handling of an initial `TypeError` during one-hot encoding by correcting the parameter from `drop_first=True` to `drop='first'` highlights the importance of precise API usage and error handling in data preprocessing.
